atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java

```diff
@@ -1637,6 +1638,62 @@ public void validateLocksOnCommitIfEmptyReadsFollowedByNonEmptyReadsIfValidation
         assertThatExceptionOfType(TransactionLockTimeoutException.class).isThrownBy(transaction::commit);
     }
 
+    @Test
+    public void skipImmutableTLocksIfReadingEqualsExpectedNumberOfValuesEvenWhenRequestedMore() {
+        putCellsInTable(List.of(TEST_CELL), TABLE_SWEPT_THOROUGH);
+
+        TimelockService spiedTimeLockService = spy(timelockService);
+        long transactionTs = spiedTimeLockService.getFreshTimestamp();
+        LockImmutableTimestampResponse res = spiedTimeLockService.lockImmutableTimestamp();
+
+        Transaction transaction = getSnapshotTransactionWith(
+                spiedTimeLockService, () -> transactionTs, res, PreCommitConditions.NO_OP, true);
+
+        transaction.getWithExpectedNumberOfCells(TABLE_SWEPT_THOROUGH, ImmutableSet.of(TEST_CELL, TEST_CELL_2), 1);
+        transaction.commit();
+        spiedTimeLockService.unlock(ImmutableSet.of(res.getLock()));
+
+        verify(spiedTimeLockService, never()).refreshLockLeases(ImmutableSet.of(res.getLock()));
+    }
+
+    @Test
+    public void performImmutableTLocksIfReadingLessThanExpectedNumberOfValues() {
+        putCellsInTable(List.of(TEST_CELL), TABLE_SWEPT_THOROUGH);
+
+        TimelockService spiedTimeLockService = spy(timelockService);
+        long transactionTs = spiedTimeLockService.getFreshTimestamp();
+        LockImmutableTimestampResponse res = spiedTimeLockService.lockImmutableTimestamp();
+
+        Transaction transaction = getSnapshotTransactionWith(
+                spiedTimeLockService, () -> transactionTs, res, PreCommitConditions.NO_OP, true);
+
+        transaction.getWithExpectedNumberOfCells(
+                TABLE_SWEPT_THOROUGH, ImmutableSet.of(TEST_CELL, TEST_CELL_2, TEST_CELL_3), 2);
+        transaction.commit();
+        spiedTimeLockService.unlock(ImmutableSet.of(res.getLock()));
+
+        verify(spiedTimeLockService, times(1)).refreshLockLeases(ImmutableSet.of(res.getLock()));
+    }
+
+    @Test
+    public void throwsIfReadingMoreThanExpectedNumberOfValues() {
+        putCellsInTable(List.of(TEST_CELL, TEST_CELL_2), TABLE_SWEPT_THOROUGH);
+
+        TimelockService spiedTimeLockService = spy(timelockService);
+        long transactionTs = spiedTimeLockService.getFreshTimestamp();
+        LockImmutableTimestampResponse res = spiedTimeLockService.lockImmutableTimestamp();
+
+        Transaction transaction = getSnapshotTransactionWith(
+                spiedTimeLockService, () -> transactionTs, res, PreCommitConditions.NO_OP, true);
+
+        assertThatThrownBy(() -> transaction.getWithExpectedNumberOfCells(
+                        TABLE_SWEPT_THOROUGH, ImmutableSet.of(TEST_CELL, TEST_CELL_2), 1))
+                .isInstanceOf(SafeIllegalStateException.class)
+                .hasMessageContaining("KeyValueService returned more results than Get expected");
+
+        verify(spiedTimeLockService, never()).refreshLockLeases(ImmutableSet.of(res.getLock()));
+    }
+
```

Comment:
question: Is it possible to add two tests for the Lock Watch cache, to ensure that we subtract from the expected number _only when_ the cached values were non-empty?

Reply:
added!